### PR TITLE
Generalize AbstractHyperrectangle -> Zonotope conversion

### LINKED ISSUE #1879: Generalize conversion from AbstractHyperrectangle to Zonotope
The `convert` implementation

https://github.com/JuliaReach/LazySets.jl/blob/85e4b30770a2d888a1b10951ed687e1ac49accbd/src/convert.jl#L258-L260

should be generalized from `AbstractHyperrectangle` to `AbstractZonotope` and rely on `genmat` instead of a custom implementation (after doing #1878 this will be equivalent).

### LINKED ISSUE #2248: Fix conversion of SA hyperrectangular sets to zonotopes
see [this comment](https://github.com/JuliaReach/LazySets.jl/pull/2214#issuecomment-659858766)

---

There is an issue with mixed array types. I guess that `_split_ret` for the two other combinations (where only one component is an `MArray`) would be needed. You can either fix it or open an issue.
We should also open an issue that in the conversion to zonotopes from static arrays we should use a static generator matrix.
```julia
julia> B = BallInf(SVector{2}(1., 1.), 1.)
BallInf{Float64,SArray{Tuple{2},Float64,1,2}}([1.0, 1.0], 1.0)

julia> convert(Zonotope, B)
Zonotope{Float64,SArray{Tuple{2},Float64,1,2},Array{Float64,2}}([1.0, 1.0], [1.0 0.0; 0.0 1.0])
# note that only the center is a static vector; the generators are a normal matrix

julia> split(B, [1], [1])
ERROR: MethodError: Cannot `convert` an object of type 
  Zonotope{Float64{},MArray{Tuple{2},Float64,1,2},Array{Float64{},2}} to an object of type 
  Zonotope{Float64{},SArray{Tuple{2},Float64,1,2},Array{Float64{},2}}
Closest candidates are:
  convert(::Type{T}, ::T) where T at essentials.jl:171
Stacktrace:
 [1] setindex!(::Array{Zonotope{Float64,SArray{Tuple{2},Float64,1,2},Array{Float64,2}},1}, ::Zonotope{Float64,MArray{Tuple{2},Float64,1,2},Array{Float64,2}}, ::Int64) at ./array.jl:826
 [2] _append! at ./array.jl:968 [inlined]
 [3] append! at ./array.jl:960 [inlined]
 [4] _split(::Zonotope{Float64,SArray{Tuple{2},Float64,1,2},Array{Float64,2}}, ::Array{Int64,1}, ::Array{Int64,1}) at .julia/dev/LazySets/src/Sets/Zonotope.jl:402
 [5] split(::BallInf{Float64,SArray{Tuple{2},Float64,1,2}}, ::Array{Int64,1}, ::Array{Int64,1}) at .julia/dev/LazySets/src/Interfaces/AbstractZonotope.jl:652
```

### PR DESCRIPTION
**Pull request by mforets**
Closes #2248.
Closes #1879.

---

This PR incorporates the change from https://github.com/JuliaReach/LazySets.jl/pull/2251.

Moreover, i added static versions for `BallInf` generators and `Hyperrectangle` generators. For the latter, i couldn't find a efficient implementation which considers the (runtime) behavior of possibly zero radius components. So i kept both versions. The reference notebook is again https://nbviewer.jupyter.org/github/mforets/escritoire/blob/master/2020/Week42/LazySets%232248.ipynb

### CODE REVIEW

**Review by schillic — APPROVED** [review]
LGTM, maybe add a test?

**Comment by mforets** [comment]
> LGTM, maybe add a test?

yes. i'll add new tests. i was a bit disappointed that this PR doesn't actually fix `genmat` in a generic way for SArrays. it doesn't seem obvious to me. maybe using generated functions.

**Comment by mforets** [comment]
updated notebook:  https://nbviewer.jupyter.org/github/mforets/escritoire/blob/master/2020/Week44/Zonotope_Hyperrect.ipynb

File: test/unit_Hyperrectangle.jl [hunk 1]
```diff
@@ -235,6 +235,34 @@ for N in [Float64, Rational{Int}, Float32]
...
+    # conversion to a zonotope
+    if N <: Rational
+        H = Hyperrectangle(low=N[5//10, 6//10], high=N[124//10, 2355//10])
+        Hz = convert(Zonotope, H)
+        @test Hz == Zonotope(N[129//20, 2361//20], N[119//20 0//1; 0//1 2349//20])
+    else
+        H = Hyperrectangle(low=N[0.5, 0.6], high=N[12.4, 235.5])
+        Hz = convert(Zonotope, H)
+        @test Hz == Zonotope(N[6.45, 118.05], N[5.95 0.0; 0.0 117.45])
+    end
```
Comment (schillic) [inline test/unit_Hyperrectangle.jl:247]:
```suggestion
    H = Hyperrectangle(low=N[5//10, 6//10], high=N[124//10, 2355//10])
    Hz = convert(Zonotope, H)
    @test Hz == Zonotope(N[129//20, 2361//20], N[119//20 0//1; 0//1 2349//20])
```
For me this works.

**Review by schillic — APPROVED** [review]

File: test/unit_Hyperrectangle.jl [hunk 2]
```diff
@@ -235,6 +235,34 @@ for N in [Float64, Rational{Int}, Float32]
...
+
+    # conversion of a hyperrectangle with static array components to a zonotope
+    # the specialized method for 2D static arrays is also tested
+    if N <: Rational
+        H = Hyperrectangle(low=SA[5//10, 6//10], high=N[124//10, 2355//10])
```
Comment (schillic) [inline test/unit_Hyperrectangle.jl:252]:
```suggestion
        H = Hyperrectangle(low=SA[5//10, 6//10], high=SA[124//10, 2355//10])
```

Comment (schillic) [reply test/unit_Hyperrectangle.jl:252]: This is the way to avoid the case split for static arrays :tada:

File: test/unit_Hyperrectangle.jl [hunk 3]
```diff
@@ -235,6 +235,34 @@ for N in [Float64, Rational{Int}, Float32]
...
+        Z = Zonotope(SA[129//20, 2361//20], SA[119//20 0//1; 0//1 2349//20])
+        @test Hz == Z && Hz_sp == Z
+
+    else
+        H = Hyperrectangle(low=SA[0.5, 0.6], high=SA[12.4, 235.5])
+        Hz = convert(Zonotope, H)
+        Hz_sp = LazySets._convert_2D_static(Zonotope, H)
+        Z = Zonotope(SA[6.45, 118.05], SA[5.95 0.0; 0.0 117.45])
+        @test Hz == Z && H == Z
+    end
```
Comment (schillic) [inline test/unit_Hyperrectangle.jl:264]:
```suggestion
    H = Hyperrectangle(low=SA[N(5//10), N(6//10)], high=N[N(124//10), N(2355//10)])
    Hz = convert(Zonotope, H)
    Hz_sp = LazySets._convert_2D_static(Zonotope, H)
    Z = Zonotope(SA[N(129//20), N(2361//20)], SA[N(119//20) N(0//1); N(0//1) N(2349//20)])
    @test Hz == Z && Hz_sp == Z
```

**Review by schillic — APPROVED** [review]

**Comment by schillic** [comment]
build error: `no docs found for 'convert(::Type{Zonotope}, ::AbstractHyperrectangle{N}) where {N}'`
(You removed the docstring, so you also need to remove it from the `docs` folder.)

**Pull request merged**